### PR TITLE
Change BlockReader API

### PR DESCRIPTION
**Pull request by hmacdope**
This is an updated version of the CythonBlockReader API, incorporating some changes that @richardjgowers  suggested.

- Primary change is that you now need to provide a numpy array to read block data into rather than it being tied to the block instance.

Still not quite complete as I need to improve the array copying and add some memoryviews.

### CODE REVIEW

**Comment by hmacdope** [comment]
ready for review :)

**Comment by orbeckst** [comment]
@hmacdope impressive progress. Sorry, my review was hastily written and does not convey how happy I am that this appears to be at an usable state.

**Comment by hmacdope** [comment]
Thanks for the feedback @orbeckst, I'll work through bit by bit over coming days.

File: pytng/pytng.pyx [hunk 1]
```diff
@@ -1037,314 +1136,275 @@ cdef class TNGFileIterator:
...
+            #         data[i, j] = (<double*>values)[i * n_values_per_frame + j]
+
+        else:
+            printf("PYTNG WARNING: block datatype not understood")
+            return TNG_CRITICAL
+        
+        #NOTE
+        # couldn't get this to work by assigining the memview to data directly,
+        # seems to require and explicit copy to the destination array
+        np.copyto(data, data_loc, casting="no")
```
[the remark this answers is not in the document]
Comment (hmacdope) [reply pytng/pytng.pyx:1327]: Correct, perhaps @tylerjereddy would know If this can be done better?

Other options are:
* tying the C array to a `MemoryWrapper` instance and then using PyArray_SetBaseObject, meaning that we do the memory management for the caller (IMO bad idea).
* Iterate over the C array and put it in the NumPy array (FLOP inefficent)
* ??

File: pytng/pytng.pyx [hunk 2]
```diff
@@ -1037,314 +1136,275 @@ cdef class TNGFileIterator:
...
+                "PYTNG ERROR: Second axis must be n_values_per_frame long")
+        
+        #local array that is moved into other 
+        cdef np.ndarray data_loc
+
+        if datatype == TNG_FLOAT_DATA:  # TODO fix this to be more efficent
+            if dtype != np.float32:
+                printf(
+                    "PYTNG ERROR: dtype of array does not match tng dtype\n")
+                return TNG_CRITICAL
```
[the remark this answers is not in the document]
Comment (hmacdope) [reply pytng/pytng.pyx:1287]: My bad, hangover from this being nogil. Ill fix.

File: pytng/pytng.pyx [hunk 3]
```diff
@@ -1037,314 +1136,275 @@ cdef class TNGFileIterator:
...
+                                                    & precision,
+                                                    & datatype,
+                                                    self.debug)
+
+        if read_stat != TNG_SUCCESS:
+            data = None
+            return
+
+        if data.ndim > 2:
+            raise IndexError("PYTNG ERROR: Numpy array must be 2 dimensional")
```
[the remark this answers is not in the document]
Comment (hmacdope) [reply pytng/pytng.pyx:1270]: Yes as we have GIL here. :)

File: pytng/pytng.pyx [hunk 4]
```diff
@@ -1037,314 +1137,301 @@ cdef class TNGFileIterator:
...
-            2, dims, NPY_DOUBLE, self._wrapper.ptr)
-        Py_INCREF(self._wrapper)
-        err = PyArray_SetBaseObject(self.values, self._wrapper)
-        if err:
-            raise ValueError("Array object cannot be created")
-        if self.debug:
-            print(self.values)
+        cdef np.float32_t[:, :] _float_view
+        cdef int64_t[:, :] _int64_t_view
+        cdef double[:, :] _double_view
```
Comment (tylerjereddy) [inline pytng/pytng.pyx:1254]: Are any of these typed memoryviews going to hold C-contiguous data? You've probably thought of that already, but can maybe use the `::1` notation to squeeze out a little more perf under the hood.

File: docs/index.rst [hunk 5]
```diff
@@ -10,51 +10,61 @@
...
+  import numpy as np
 
-  with pytng.TNGFile('traj.tng', 'r') as f:
-      first_frame = f[0]
-      last_frame = f[-1]
-
-      every_other_frame = [ts for ts in f[::2]]
+  with pytng.TNGFileIterator('traj.tng', 'r') as tng:
+      tng[0].current_integrator_step.get_pos(positions)
+      tng[0,100,10].get_pos(positions)
```
[the remark this answers is not in the document]
Comment (hmacdope) [reply docs/index.rst:59]: gone now as misleading and will only return last frame.

File: docs/index.rst [hunk 6]
```diff
@@ -10,51 +10,61 @@
...
-For example, the coordinate information can be accessed via the `.positions` 
-attribute of the returned object.
 
 .. code-block:: python
 
-  import pytng
+   import pytng
+   import numpy as np
+
+   with pytng.TNGFileIterator("tng.tng", 'r') as tng:
```
[the remark this answers is not in the document]
Comment (hmacdope) [reply docs/index.rst]: I think `traj.tng` should be clear enough.

File: pytng/pytng.pyx [hunk 7]
```diff
@@ -14,6 +14,7 @@ from libc.stdio cimport printf, FILE, SEEK_SET, SEEK_CUR, SEEK_END
 from libc.stdlib cimport malloc, free, realloc
 from libc.stdint cimport int64_t, uint64_t, int32_t, uint32_t
 from cpython.mem cimport PyMem_Malloc, PyMem_Realloc, PyMem_Free
+# TODO REMOVE
```
[the remark this answers is not in the document]
Comment (hmacdope) [reply pytng/pytng.pyx]: There is a fair bit of junk hanging around to support the old API. Should I make a separate PR to strip it all out or clean up here?

File: pytng/pytng.pyx (hunk 4 above)
Comment (hmacdope) [reply pytng/pytng.pyx]: I think new structure resolves this, but would be grateful if you could take a quick peek. :)

File: pytng/pytng.pyx [hunk 8]
```diff
@@ -775,150 +790,228 @@ cdef class TNGFileIterator:
...
-                             "in input file {}".format(self._n_frames))
+        :rtype: :class:`TNGCurrentIntegratorStep`
+        """
+        if not self.is_open:
+            raise IOError("File is not yet open")
+        return self.current_step
 
-        self.step = frame
-        self.current_step = TNGCurrentIntegratorStep(debug=self.debug)
+    cpdef read_step(self, step):
```
[the remark this answers is not in the document]
Comment (hmacdope) [reply pytng/pytng.pyx:984]: no problem, I think that is a sensible pattern for access?

File: docs/index.rst [hunk 9]
```diff
@@ -10,51 +10,61 @@
...
+      # make a numpy array to hold the data using helper function
+      positions = tng.make_ndarray_for_block_from_name("TNG_TRAJ_POSITIONS")
+      
+      # stride over the whole trajectory for the frames that have data
+      for ts in range(0, len(tng), tng.block_strides["TNG_TRAJ_POSITIONS"]):
+         # read the integrator timestep
+         tng.read_step(ts)
+         # get the data from the requested block by supplying NumPy array
+         tng.current_integrator_step.get_pos(positions)
+
```
[the remark this answers is not in the document]
Comment (hmacdope) [reply docs/index.rst:79]: we are now both reading in place and returning the value ie:

`positions = ts.get_positions(positions)`

and 

`ts.get_positions(positions)`

both work.

Is this how we want the final product to look, or would we prefer just a single method to be canonical?

[the remark this answers is not in the document]
Comment (hmacdope) [reply docs/index.rst:79]: or do we want to rename the getter as well :)

[the remark this answers is not in the document]
Comment (orbeckst) [reply docs/index.rst:79]: I like the returning values in addition to the in-place. 

I can live with `get_*()` if you feel that this is the best name. After all, this is more your code than mine ;-).

File: docs/index.rst [hunk 10]
```diff
@@ -10,51 +10,102 @@
...
+* block :  a data element in a TNG trajectory containing a specific type of data eg. positions, velocities or box vectors.
+* block id : an integer (a long long) that indicates the type of data contained in a block.
+* block name : a name that matches a specific block id, normally starts with the TNG prefix.
+* particle dependency : indicates whether the data in a block is dependent on the particles in the simulation, (eg positions) or is not (eg box vectors).
+
+Notes on the TNG format and PyTNG
+=================================
+
+While the TNG format supports storage of simulations conducted in the
+grand canonical ensemble, PyTNG does not currently support this. Additonally,
```
Comment (orbeckst) [inline docs/index.rst:32]: Additionally

File: docs/index.rst [hunk 11]
```diff
@@ -10,51 +10,102 @@
...
+PyTNG. For more information on the original TNG API, see the following papers
+[1_] [2_].
+
+* step : an integrator timestep (one MD step).
+* frame : an integrator timestep with data associated.
+* stride : the number of *steps* between writes of data to the trajectory.
+* block :  a data element in a TNG trajectory containing a specific type of data eg. positions, velocities or box vectors.
+* block id : an integer (a long long) that indicates the type of data contained in a block.
+* block name : a name that matches a specific block id, normally starts with the TNG prefix.
+* particle dependency : indicates whether the data in a block is dependent on the particles in the simulation, (eg positions) or is not (eg box vectors).
```
Comment (orbeckst) [inline docs/index.rst:26]: e.g.,

File: docs/index.rst [hunk 12]
```diff
@@ -10,51 +10,102 @@
...
+grand canonical ensemble, PyTNG does not currently support this. Additonally,
+the TNG format includes a TNG_MOLECULES block that contains the simulation
+topology. PyTNG does not currently make use of this information.
 
-TNG files can be read using the TNGFile class as a file handle,
+
+Usage example for TNGFileIterator
+=================================
+
+TNG files can be read using the TNGFileIterator class as a file handle,
```
Comment (orbeckst) [inline docs/index.rst:40]: link 
```reST
:class:`TNGFileIterator`
```

File: docs/index.rst [hunk 13]
```diff
@@ -10,51 +10,102 @@
...
+      # make a numpy array to hold the data using helper function
+      # this array will then be updated in-place 
+      
+      positions = tng.make_ndarray_for_block_from_name("TNG_TRAJ_POSITIONS")
+      
+      # the TNG API uses regular strides for data deposition, here we stride
+      # over the whole trajectory for the frames that have position data
+      # len(tng) is the total number of steps in the file.
+
+      for ts in tng[0:len(tng):tng.block_strides["TNG_TRAJ_POSITIONS"]]:
```
Comment (orbeckst) [inline docs/index.rst:71]: Just a question: is there a way to iterate over all steps that contain any data, e.g., if block_strides for positions are 100 and 1000 for velocities then we would automatically do stride 100 and for 9 out of 10 steps, we just don't get velocity data.

More weirdly, when positions are saved every 5 steps and velocities every 11 steps then we would get only a few steps that have both (at step = n * 5 * 11, n = 0, 1, ,2, ...).

File: docs/index.rst [hunk 14]
```diff
@@ -10,51 +10,102 @@
...
 
-      every_other_frame = [ts for ts in f[::2]]
+
+Error handling
+==============
+
+PyTNG is designed so that Python level exceptions can (and should) be caught by
+the calling code. However, as the trajectory reading itself is done 
+by Cython at the C level with the GIL released, low level trajectory reading
+errors are caught by more general exceptions at the Python level.
```
Comment (orbeckst) [inline docs/index.rst:100]: Are there any specific exceptions that the user needs to be aware of?

Are exceptions documented in the classes (sorry, I am short on time reviewing this)

File: pytng/pytng.pyx [hunk 15]
```diff
@@ -768,157 +770,267 @@ cdef class TNGFileIterator:
...
-        if self.current_step.block_set.get(TNG_TRAJ_FORCES) is None:
-            return None
-        else:
-            return self.current_step.block_set.get(TNG_TRAJ_FORCES).values
+        Returns
+        -------
+        values_per_frame : dict
+            dictionary of block names (keys) and number of values per frame
+            (values)
+        :rtype: dict
```
Comment (orbeckst) [inline pytng/pytng.pyx:897]: I think you're mixing doc styles here. I the `:rtype` line is superfluous.

File: pytng/pytng.pyx [hunk 16]
```diff
@@ -1016,335 +1133,328 @@ cdef class TNGFileIterator:
...
+        datatype : char *
+            set to the datatype of the block
+        debug : bint
+            debug the block read
+
+        Notes
+        -----
+        This function is marked nogil and is called without the GIL so cannot
+        contain python or python exceptions. Instead failure is marked by
+        returning TNG_CRITICAL. Cleanup must then be done by the calling code.
```
Comment (orbeckst) [inline pytng/pytng.pyx:1387]: Is this supposed to be 
```reST
:data:`TNG_CRITICAL`
```
Not clear from the docs how the user can check for the status code.

File: pytng/pytng.pyx [hunk 17]
```diff
@@ -597,9 +594,9 @@ cdef extern from "tng/tng_io.h":
         int64_t * n_values_per_frame,
         char * type) nogil
 
+# TODO REMOVE
```
Comment (orbeckst) [inline pytng/pytng.pyx:597]: remove? is the frame tuple still used?

File: pytng/pytng.pyx [hunk 18]
```diff
@@ -1016,335 +1133,328 @@ cdef class TNGFileIterator:
...
-            printf("PYTNG WARNING: char data reading is not implemented \n")
+    cdef tng_function_status _get_step_time(self, double * step_time):
+        stat = tng_util_time_of_frame_get(self._traj, self.step, step_time)
+        if stat != TNG_SUCCESS:
+            return TNG_CRITICAL
 
-        else:
-            printf("PYTNG WARNING: block data type %d not understood \n", datatype)
 
+global block_dictionary
```
Comment (orbeckst) [inline pytng/pytng.pyx:1456]: globals... really?

File: docs/index.rst [hunk 19]
```diff
@@ -10,51 +10,102 @@
...
+  import numpy as np
 
-  with pytng.TNGFile('traj.tng', 'r') as f:
-      first_frame = f[0]
-      last_frame = f[-1]
+  with pytng.TNGFileIterator('traj.tng', 'r') as tng:
+      tng[100].get_positions(positions)
 
-      every_other_frame = [ts for ts in f[::2]]
+
```
Comment (orbeckst) [inline docs/index.rst:93]: What other data can be read? Explain velocity and forces.

File: docs/index.rst [hunk 20]
```diff
@@ -10,51 +10,102 @@
...
+      # over the whole trajectory for the frames that have position data
+      # len(tng) is the total number of steps in the file.
+
+      for ts in tng[0:len(tng):tng.block_strides["TNG_TRAJ_POSITIONS"]]:
+         # read the integrator timestep, modifying the current_integrator_step
+         # which is returned as ts
+
+         # get the data from the requested block by supplying NumPy array which
+         # is updated in-place or returned 
+
```
Comment (orbeckst) [inline docs/index.rst:77]: Include box vectors in example.

File: docs/index.rst (hunk 19 above)
Comment (orbeckst) [reply docs/index.rst:93]: What happens if you go to a integrator step that does not contain data? How to deal with this case?

File: docs/index.rst [hunk 21]
```diff
@@ -10,51 +10,102 @@
...
-          ts.positions
+           
 
 It is also possible to slice and index the file object to select particular
 frames
 
 .. code-block:: python
 
   import pytng
+  import numpy as np
```
Comment (orbeckst) [inline docs/index.rst:88]: Why numpy here? You're not using it (and presumably assume that `positions` is already allocated)

File: docs/index.rst (hunk 13 above)
Comment (hmacdope) [reply docs/index.rst:71]: The easiest solution I thought of was to stride over the greatest common divisor of the strides.
This is already computed as `TngFileIterator._gcd`, but I haven't made use of it yet.

Does that sound reasonable?

Comment (orbeckst) [reply docs/index.rst:71]: The GCD(5, 11) = 1 so what does this mean in terms of iterating? We get a lot of empty steps?

Comment (orbeckst) [reply docs/index.rst:71]: For the likely more common case of 10,100 the GCD sounds pretty sane, i.e., when GCD(m, n) == m or n. In the case where we get empty steps then the calling code will need to decide what to do (such as not returning a step) and that will be rather costly.

I'd much prefer if TNGFileIterator never returned empty frames. It tags frames so user code will just have to work with the frame metadata.

From MDAnalysis' perspective I also want to know what `len(tng)` should mean; I think the only sensible interpretation is "total number of non-empty frames" (like TRR).

Any opinions @richardjgowers @jbarnoud @kain88-de ?

File: pytng/pytng.pyx (hunk 17 above)
Comment (hmacdope) [reply pytng/pytng.pyx:597]: I will remove the old API once we get this one in I think?

File: docs/index.rst (hunk 13 above)
Comment (hmacdope) [reply docs/index.rst:76]: Working with steps is closer to the way the format itself works. My suggestion is that working in the original units of the format is better, we can then provide the user the tools they need to iterate with the minimum of wasted steps (obviously with big improvements in docs required).

I am happy to change it up though! No dramas.

I see where your coming from in regards to the length, would this mean the total number of frames with any block attached would become the length? This would still result in blank reads although a lot fewer. The other option is to have a separate length in each block "dimension".

Open to suggestion!

File: docs/index.rst (hunk 19 above)
Comment (hmacdope) [reply docs/index.rst:119]: see discussion above

File: docs/index.rst (hunk 13 above)
Comment (hmacdope) [reply docs/index.rst:87]: Hey @orbeckst I have tried to clear this up a bit and add more info in the docs.

For blank reads we currently fill the array with np.nan. Not the most elegant solution but gets the point across.  I have added an example that demonstrates this to the docs.

File: docs/index.rst [hunk 22]
```diff
@@ -10,51 +10,183 @@
...
+
+Notes on the TNG format and PyTNG
+=================================
+
+The TNG format is highly flexible, allowing the storage of almost any datatype
+from any point in a simulation. This information is written at certain strides,
+i.e at every *N* steps. Under most circumstances, strides are of a similar
+magnitude or share a large *greatest common divisor (GCD)*. However this is not
+always the case and introduces additional complexity in iterating through the
+file effectivley.  The main challenge is if you want to retrieve multiple datatypes in a single
```
Comment (tylerjereddy) [inline docs/index.rst:38]: typo "effectivley"

File: docs/index.rst [hunk 23]
```diff
@@ -10,51 +10,183 @@
...
+pass that do not share a large GCD in their strides, nessecitating lots of
+blank reads. Avoiding this is still a work in progress for PyTNG.
+
+While the TNG format supports storage of simulations conducted in the
+grand canonical ensemble, PyTNG does not currently support this. Additionally,
```
Comment (tylerjereddy) [inline docs/index.rst:43]: In simpler terms this means that PyTNG does not support trajectories where the number of particles varies between frames?

File: docs/index.rst [hunk 24]
```diff
@@ -10,51 +10,183 @@
...
+the TNG format includes a TNG_MOLECULES block that contains the simulation
+topology. PyTNG does not currently make use of this information.
```
Comment (tylerjereddy) [inline docs/index.rst:45]: One of the selling points of the TNG format when it was originally announced was that we wouldn't have to worry about keeping track of separate topology and trajectory files. Is the plan for MDAnalysis to eventually support a single TNG file as the source of both topology and trajectory?

File: tests/test_tng.py [hunk 25]
```diff
@@ -121,88 +123,109 @@ def test_tng_example_getitem_bool_TypeError(cls, TNG_EXAMPLE):
...
-@pytest.mark.skip(reason="FAILING")
+
 def test_tng_example_natoms(TNG_EXAMPLE_DATA, TNG_EXAMPLE):
     with pytng.TNGFileIterator(TNG_EXAMPLE) as tng:
         assert TNG_EXAMPLE_DATA.natoms == tng.n_atoms
 
+
+def test_README_example(TNG_EXAMPLE):
+    with pytng.TNGFileIterator(TNG_EXAMPLE, 'r') as tng:
+        positions = np.ndarray((tng.n_atoms, 3), dtype=np.float32)
```
Comment (tylerjereddy) [inline tests/test_tng.py:134]: `np.empty` vs. the low-level instantiator?

File: tests/test_tng.py [hunk 26]
```diff
@@ -121,88 +123,109 @@ def test_tng_example_getitem_bool_TypeError(cls, TNG_EXAMPLE):
...
 def test_tng_example_tng_example_first_positions(
     TNG_EXAMPLE_DATA, TNG_EXAMPLE
 ):
     with pytng.TNGFileIterator(TNG_EXAMPLE) as tng:
-        print(tng.block_strides)
-        first_frame = tng[0].pos
-        print(first_frame)
-        print(TNG_EXAMPLE_DATA.first_frame)
-        assert np.array_equal(TNG_EXAMPLE_DATA.first_frame, first_frame)
+        pos = np.zeros((15, 3), dtype=np.float32)
```
Comment (tylerjereddy) [inline tests/test_tng.py:151]: ok, so it looks like you can use the high-level NumPy array instantiators just fine then

File: tests/test_tng.py [hunk 27]
```diff
@@ -121,88 +123,109 @@ def test_tng_example_getitem_bool_TypeError(cls, TNG_EXAMPLE):
...
+        assert np.array_equal(TNG_EXAMPLE_DATA.first_frame, pos)
+
 
 def test_tng_example_tng_example_last_positions(TNG_EXAMPLE_DATA, TNG_EXAMPLE):
     with pytng.TNGFileIterator(TNG_EXAMPLE) as tng:
-        
-        last_frame = tng[len(tng)-1].pos
-        assert np.array_equal(TNG_EXAMPLE_DATA.last_frame, last_frame)
+        pos = np.zeros((15, 3), dtype=np.float32)
+        pos = tng[9].get_positions(pos)
```
Comment (tylerjereddy) [inline tests/test_tng.py:159]: is it worth parametrizing the indices tested to include a `-1` for reverse index support to access the last element?

File: tests/test_tng.py [hunk 28]
```diff
@@ -121,88 +123,109 @@ def test_tng_example_getitem_bool_TypeError(cls, TNG_EXAMPLE):
...
+        assert np.array_equal(TNG_EXAMPLE_DATA.last_frame, pos)
+
+
+def test_tng_example_first_and_last_seq_positions(TNG_EXAMPLE_DATA,
+ TNG_EXAMPLE):
+    with pytng.TNGFileIterator(TNG_EXAMPLE) as tng:
+        pos = np.zeros((15, 3), dtype=np.float32)
+        pos = tng[0].get_positions(pos)
+        assert np.array_equal(TNG_EXAMPLE_DATA.first_frame, pos)
+        pos = tng[9].get_positions(pos)
```
Comment (tylerjereddy) [inline tests/test_tng.py:169]: will `-1`, `-9` indices work above as well?

File: tests/test_tng.py [hunk 29]
```diff
@@ -283,42 +315,49 @@ def test_water_npt_uncompressed_vels_forces_first_vels(
...
             decimal=2,
         )  # decimal = 2 really slack
 
+
 def test_water_npt_uncompressed_vels_forces_last_frc(
     WATER_NPT_UNCOMPRESSED_VELS_FORCES, WATER_NPT_UNCOMPRESSED_VELS_FORCES_DATA
 ):
     with pytng.TNGFileIterator(WATER_NPT_UNCOMPRESSED_VELS_FORCES) as tng:
-        last_frame_last_10_frc = tng[len(tng) -1].frc[2690:2700, :]
+        frc = np.zeros((2700, 3), dtype=np.float32)
```
Comment (tylerjereddy) [inline tests/test_tng.py:357]: Is there a test for handling "invalid" `dtype` for the array used to store the data? If you feed in `np.complex128` or a unicode type or something?

File: pytng/pytng.pyx [hunk 30]
```diff
@@ -1016,334 +1138,340 @@ cdef class TNGFileIterator:
...
+        NumPy array
 
-    def __cinit__(self, TrajectoryWrapper traj, int64_t frame,
-                  bint debug=False):
+        Parameters
+        ----------
+        data : np.ndarray
+           NumPy array to read the data into. As this is a particle dependent
+           block, the shape should be (n_atoms, n_values_per_frame)
+           ie (n_atoms, 3).
```
Comment (tylerjereddy) [inline pytng/pytng.pyx:1216]: To clarify, when it says the shape should be `(n_atoms, n_values_per_frame)` is that different from `(n_atoms, ndim)`? I think when I was reading through the examples at the start I was under the impression this was the number of dims? Maybe it is meant to be the same thing.

File: pytng/pytng.pyx [hunk 31]
```diff
@@ -1016,334 +1138,340 @@ cdef class TNGFileIterator:
...
-        self._wrapper = MemoryWrapper(1)
+    cpdef np.ndarray get_velocities(self, np.ndarray data):
+        """Get the velocities present at the current step and read them into a
+        NumPy array
 
-    def __dealloc__(self):
-        pass
+        Parameters
+        ----------
+        data : np.ndarray
```
Comment (tylerjereddy) [inline pytng/pytng.pyx:1245]: As I was looking through the tests I was wondering what the supported dtypes for these input arrays might be, and if an exception would be raised for i.e., non-numeric dtypes, etc.

Looking farther down below--is this all routed through the dtype check in `get_blockid()` then?

**Review by tylerjereddy — COMMENTED** [review]
I added a few minor comments/questions since my review was requested again, but none of the comments are intended to be blocking for the merge; just suggestions or things I maybe didn't get.

I see that codecov is adding some tags for Cython lines that are not covered by tests yet; on the other hand, many foundational libraries don't even check the coverage of their Cython lines so not sure we have to be super strict on that, at least initially because this is a lot of work.

File: pytng/pytng.pyx (hunk 4 above)
Comment (tylerjereddy) [reply pytng/pytng.pyx]: It certainly looks cleaner than the hard copy from before; I'll mark as resolved.

File: README.rst [hunk 32]
```diff
@@ -13,17 +13,21 @@ pytng - A python library to read TNG files!
...
   import pytng
+  import numpy as np
 
-  with pytng.TNGFile('traj.tng', 'r') as f:
-      for ts in f:
-          time = ts.time
-          coordinates = ts.positions
+  with pytng.TNGFileIterator('traj.tng', 'r') as tng:
+
+    positions = np.ndarray((tng.n_atoms,3), dtype=np.float32)
```
[the remark this answers is not in the document]
Comment (hmacdope) [reply README.rst:26]: Ok noted!

File: docs/index.rst (hunk 24 above)
Comment (hmacdope) [reply docs/index.rst:45]: For sure thats the way forward. Issue raised to this effect in #44.

The file contains the whole gmx mtop, so would probably be able to leverage or adapt some existing MDA code.

File: pytng/pytng.pyx (hunk 31 above)
Comment (hmacdope) [reply pytng/pytng.pyx:1245]: I have tried to clarify available dtypes in the docs.

Yes you are right it all passes through that dtype check, and an exception is raised if incorrect.

**Comment by hmacdope** [comment]
Thanks for the review @tylerjereddy. Great points I will address ASAP.

File: tests/test_tng.py (hunk 27 above)
Comment (hmacdope) [reply tests/test_tng.py:176]: Negative indices not currently supported. I should add them in.

Is there a canonical way to do that with an Iterator or should I implement my own version?

File: pytng/pytng.pyx (hunk 30 above)
Comment (hmacdope) [reply pytng/pytng.pyx:1216]: I think meant to be the same thing.

I think name `n_values_per_frame` (used in the library itself) is meant to provide the idea that for non-particle dependent blocks e.g box vectors we should expect that many values.

Open to changing if you think it would be clearer :)

File: tests/test_tng.py (hunk 28 above)
Comment (hmacdope) [reply tests/test_tng.py:197]: now they will.:)

File: tests/test_tng.py (hunk 27 above)
Comment (hmacdope) [reply tests/test_tng.py:176]: Fixed for proper wraparound on the iterator.

File: pytng/pytng.pyx (hunk 17 above)
Comment (orbeckst) [reply pytng/pytng.pyx:597]: Do you mean it should be removed as part of this PR or that once this PR is in you can work on other things to eventually remove it?

File: docs/index.rst (hunk 13 above)
Comment (orbeckst) [reply docs/index.rst:95]: Is processing np.nan expensive? 

In any case, I don't have a better idea right away (I guess we can't temporarily set the whole array to NULL or None?) Is there a way for calling code to quickly learn what data are available – is it just `np.all(np.isnan())`, which would be expensive? Could you add an example that shows how calling code can find out efficiently which data it can pull out?

Comment (orbeckst) [reply docs/index.rst:95]: The example above is for the simple case when the strides coincide. That's good to introduce the basics.

Now add an example where the strides differ. If I don't know how to do this I have no idea how write a Reader for e.g. MDA.

File: docs/index.rst [hunk 33]
```diff
@@ -75,20 +99,85 @@ An example of how to read positions from a TNG file is shown below.
...
          # is updated in-place or returned 
          
-         positons = ts.get_positions(positions)
+         # update in place
+         ts.get_positions(positions)
+         # positions = ts.get_positions(positions) is equivalent
+
+         # or return by value
+         box_vec = ts.get_box(box_vec)
+         # ts.get_box(box_vec) is equivalent
```
Comment (orbeckst) [inline docs/index.rst:108]: Stay consistent... I was confused "What, this one returns a value and the other does in place?" until I read the comment. Just make both code examples "in place" and then show in the comments the return value form.

File: docs/index.rst [hunk 34]
```diff
@@ -75,20 +99,85 @@ An example of how to read positions from a TNG file is shown below.
...
+      step = 42
+
+      # check that we are off stride (stride%step != 0)
+      assert(tng.block_strides["TNG_TRAJ_POSITIONS"]%step != 0 )
+
+      # get the data, which will be returned full of np.nan
+      tng[step].get_positions(positions)
+
+      # check that the read was blank
+      is_blank_read = np.all(np.isnan(positions)) # this will be true
```
Comment (orbeckst) [inline docs/index.rst:149]: This is the example that I asked for in an earlier comment (when I hadn't seen this one). 

The question remains how expensive `np.all(np.isnan(positions))` is. Don't the TNG iterator "know" that it filled the array and couldn't it set a flag that could be queried by calling code? Then we wouldn't even have to fill the array, calling code would just have to check the flag.

File: tests/test_tng.py (hunk 28 above)
Comment (orbeckst) [reply tests/test_tng.py:197]: Add documentation, e.g. saying that standard Python slices are supported.

Are `slice()` objects also supported?

**Comment by orbeckst** [comment]
@hmacdope I added a few more hastily written comments. Coming along nicely, I don't want to overly hold up proceedings so please feel free to push for moving forward and/or suggesting to move forward with as is and raise other issues as necessary.

File: pytng/pytng.pyx (hunk 17 above)
Comment (hmacdope) [reply pytng/pytng.pyx:597]: Sorry should have been clearer.

I'll remove the old API in a separate PR once this PR is merged. There is a lot of stuff that can be cleaned up but we have to remove it all at once so it can all still compile.

Comment (orbeckst) [reply pytng/pytng.pyx:597]: Thank you!

File: docs/index.rst (hunk 34 above)
Comment (hmacdope) [reply docs/index.rst:149]: I can add a query-able flag on the integrator step, or alternatively we can possible return a `NamedTuple` with the read status contained as one of the elements.

 I'll play with a few options :)

Comment (hmacdope) [reply docs/index.rst:163]: There is now a query-able flag for the success of the last read in `CurrentIntegratorStep.read_success`. Open to other suggestions :)

I think we should still fill the target array with `np.nan` for blank reads, otherwise the current contents of the target array will be the last read step which is potentially misleading when iterating.

I'll do some timings to see how costly filling with `np.nan` is.

File: docs/index.rst [hunk 35]
```diff
@@ -183,6 +197,10 @@ method, where the block id needs to be supplied and can be accessed from the
 Error handling
 ==============
 
+The success of the last call to `get_blockid`, `get_positions`, `get_box` and
+related methods can be checked using the :class:`TNGCurrentIntegratorStep.read_success`
+property, which returns `True` if the attempt to read was successful or `False` if not.
```
Comment (orbeckst) [inline docs/index.rst:202]: Explain what "successful" means. In particular, can it fail because (1) EOF, (2) file corruption, (3) no data stored for the step? Are these all equally unsuccessful?

Comment on reST: use double-backticks to format as code. Single backtick indicates a role in reST/sphinx (such as 
```reST
This indicates a class :class:`Universe` with the ":class:" role.
If I write `Universe` then it assumes the default role (whatever that is).
To explicitly mark up code use double back-ticks, e.g., ``u = mda.Universe(...)``, or ``(True != False) == True`` is ``True``.
```

and it just uses the default role

File: docs/index.rst (hunk 34 above)
Comment (orbeckst) [reply docs/index.rst:163]: I'd be curious about the penalty for filling the array. It seems wasteful to do.

Comment (orbeckst) [reply docs/index.rst:163]: In MDA we have `has_positions`, `has_velocities`, `has_forces` flags on `Timesteps`. Would it make sense to have a bit array for all content in TNG trajectory that, for each queried step, just has a True or False for content present?

File: docs/index.rst (hunk 35 above)
Comment (hmacdope) [reply docs/index.rst:202]: sorry for delay, will get to this asap.

**Comment by orbeckst** [comment]
@hmacdope please ping me when you need me to look at things — this seems so far along that it's already in a superposition of merged and umerged ("Schrödinger's PR").

File: docs/index.rst (hunk 34 above)
Comment (hmacdope) [reply docs/index.rst:163]: Makes sense and could be part of enhancements surrounding iteration. I'll raise an issue.

Comment (hmacdope) [reply docs/index.rst:163]: see #47

**Comment by hmacdope** [comment]
Hey @orbeckst, ready for re-review (I think).
I have added in some more explanation about calls to `read_success`.

I think we should possibly press ahead with this PR and add new features iterativley? I have raised a few new issues that we can address.

File: docs/index.rst [hunk 36]
```diff
@@ -197,15 +197,16 @@ method, where the block id needs to be supplied and can be accessed from the
...
-
 PyTNG is designed so that Python level exceptions can (and should) be caught by
 the calling code. However, as the trajectory reading itself is done 
 by Cython at the C level with the GIL released, low level trajectory reading
 errors are caught by more general exceptions at the Python level.
 
+The success of the last call to ``get_blockid``, ``get_positions``, ``get_box`` and
+related methods can be checked using the :class:`TNGCurrentIntegratorStep.read_success`
+property, which returns ``True`` if the attempt to read was successful or ``False`` if not.
+Success (``True``) inidcates that there was data of the requested block type present at the current step, while failure (``False``) indicates either no data present, file corruption or a file reading excepetion not caught by PyTNG.
```
Comment (orbeckst) [inline docs/index.rst:208]: long line and "exception", "indicates" (spell check ;-) )

**Review by orbeckst — APPROVED** [review]
I am ok with merging this PR and forging ahead.

I found two typos but that's not a blocking request... I am not a big fan of having ``False`` for "no data" and "file corruption", "end of file" etc but for right now it's consistent with the underlying library and we could improve later.

Thanks for raising the follow-up issue.

**Comment by orbeckst** [comment]
@hmacdope are you able to merge the PR yourself once you feel it's ready?

**Comment by hmacdope** [comment]
@orbeckst sure am. I'll fix those typos.
